src/properties.c

```diff
@@ -0,0 +1,683 @@
+	GtkWidget *search_entry;
+	GtkWidget *search_warning;
+
+	GtkWidget *main_box;
+	GtkWidget *left_column;
+	GtkWidget *right_column;
+	GList *field_list;
+	int field_list_length;
+	gboolean ignore_case;
+};
+
+/* This macro defines the Properties type.
+ * (GObject boilerplate)
+ */
+G_DEFINE_TYPE( Properties, properties, GTK_TYPE_WIDGET );
+
+/* This enum defines the signals used to get and set custom properties.
+ * See:
+ * 	- properties_get_property
+ * 	- properties_set_property
+ *
+ * (GObject boilerplate).
+ */
+enum {
+	PROP_IMAGE_WINDOW = 1,
+	PROP_REVEALED,
+
+	SIG_LAST
+};
+
+/* Clean up any existing children of @left_column and @right_column and any
+ * existing elements of the GList @field_list.
+ *
+ * @m	Properties *	this
+ */
+static void
+properties_clear_main_box( Properties *p )
+{
+	GtkWidget *t0, *t1;
+
+	t0 = gtk_widget_get_last_child( p->left_column );
+	while ( t0 ) {
+		t1 = gtk_widget_get_prev_sibling( t0 );
+		properties_util_free_label_box( t0 );
+		t0 = t1;
+	}
+
+	t0 = gtk_widget_get_last_child( p->right_column );
+	while ( t0 ) {
+		t1 = gtk_widget_get_prev_sibling( t0 );
+		properties_util_free_input_box( t0 );
+		t0 = t1;
+	}
+
+	if( p->field_list ) {
+		g_list_free( p->field_list );
+		p->field_list = NULL;
+		p->field_list_length = 0;
+	}
+}
+
+/* Create a new main_box. The caller should use properties_clear_main_box to clean
+ * up existing input widgets first.
+ *
+ * @m	Properties *	this
+ */
+static void
+properties_create_main_box( Properties *p )
+{
+	TileSource *tile_source;
+	GtkWidget *t;
+	VipsImage *image;
+	gchar **fields, *field;
+
+	/* If there is no TileSource, clean up any existing main_box and do
+	 * nothing.
+	 */
+	if( !(tile_source = image_window_get_tile_source( p->image_window )) ) {
+		properties_clear_main_box( p );
+		return;
+	}
+
+	/* If there is a TileSource, get its VipsImage @image and continue.
+	 */
+	image = tile_source_get_image( tile_source );
+
+	/* Clean up the old GtkGrid @p->main_box and GList @p->field_list. If
+	 * there is nothing to clean up, nothing happens.
+	 */
+	properties_clear_main_box( p );
+
+	/* Get the array of image property name character arrays. We are
+	 * responsible for freeing this with g_strfreev. The Match methods do
+	 * not modify/free the character arrays in @fields, so we only free them
+	 * once the relevant data has been copied into widgets with their
+	 * character array setter methods.
+	 */
+	fields = vips_image_get_fields( image );
+
+	/* Make new GList @p->field_list using the fields from the VipsImage
+	 * @image.
+	 */
+	p->field_list = NULL;
+	for ( int i=0; (field = fields[i]); i++ )
+		p->field_list = g_list_append( p->field_list, field );
+	p->field_list_length = g_list_length( p->field_list );
+
+	/* Add @p->field_list_length rows and 2 columns to the main_box. There
+	 * is a row for every field, i.e., every property of the VipsImage
+	 * @image.
+	 */
+	for ( int i = 0; (field = g_list_nth_data( p->field_list, i )); i++ ) {
```

Comment:
This is a bit inefficient -- `g_list_nth_data()` will loop over the whole field list each time. How about eg.

```C
for (int i = 0, GList *f = p->field_list; p; p = p->next, i++) {
    const char *field = (const char *) f->data;

    ...
}
```